### DISCUSSION TITLE
How to compile google filament to wasm for windows?

### DISCUSSION THREAD

**Discussion by amarav** [google/filament#3518] [General]
I would like to know how to compile the filament to wasm in Windows. The readme describes only for linux / macOs. Could anyone help with this?

**Comment by cx20** [comment]
@amarav This is not the way to build it, but you can get the Nightly Build version of wasm from

https://github.com/google/filament/actions?query=workflow%3AWeb

![image](https://user-images.githubusercontent.com/1833979/107109140-f1fe2780-6880-11eb-8350-2aacce67ec17.png)

![image](https://user-images.githubusercontent.com/1833979/107109144-fa566280-6880-11eb-9256-3bda7c6dc527.png)

**Comment by amarav** [comment]
@cx20 Thanku so much for sharing! I tried to view the sample https://github.com/google/filament/blob/main/web/docs/tutorial_triangle.md in browser but cant see anything... Could you share a simple program which uses this for me to understand the flow? Sorry im new to these,,,

**Reply by cx20** [reply]
Currently, Triangle samples do not work on `Chrome 88` in `Filament v1.9.10`.
It will be fixed in `Filament v1.9.11`, but if you want to try it right away, try Firefox.

The following is the display result in `Firefox 85`.
https://google.github.io/filament/webgl/tutorial_triangle.html
![image](https://user-images.githubusercontent.com/1833979/107111387-1cf17700-6893-11eb-9a34-2327f565b7d4.png)

**Comment by amarav** [comment]
@cx20 Can you share some inputs on how the canvas is allocated on the browser and how the filament apis are bind with JS using embind?

**Reply by cx20** [reply]
If you want to run the Triangle sample in Chrome right now, try the attached [zip](https://github.com/google/filament/files/5936412/triangle.zip).
Includes `Filament v1.9.11rc` and samples.

[triangle.zip](https://github.com/google/filament/files/5936412/triangle.zip)

```
\---triangle
        filament.js
        filament.wasm
        index.html
        nonlit.filamat
        nonlit.mat
```
![image](https://user-images.githubusercontent.com/1833979/107111629-f03e5f00-6894-11eb-81e6-4d00161050df.png)

**Reply by amarav** [reply]
@cx20 Thanks for your quick inputs! Im now able to run it.
What I would like to know about the source of filament https://github.com/google/filament/blob/9f68daeb8f18b8d13e66105a009e59644531a8a2/web/filament-js/extensions.js#L56

Like how the canvas is allocated and embind with the apis ?

**Reply by cx20** [reply]
See below for how to connect a Canvas to an Engine.
https://github.com/google/filament/blob/main/web/docs/tutorial_triangle.md#create-the-engine-and-scene

![image](https://user-images.githubusercontent.com/1833979/107112361-63969f80-689a-11eb-9c43-127809e1c204.png)

**Reply by amarav** [reply]
@cx20 Thanks a million! I would like to know if there is a way to read a file in wasm. In windows, we have to provide the full path. so i pass base path in resources' path. The paths are dynamic depending on project. Can you share some insights on this?

**Reply by cx20** [reply]
`filament.wasm` refers to the same location as the path specified in `filament.js`.

If no path is specified, the current path is referenced.
```html
<script src="filament.js"></script>
```

If a path is specified, its location will be referenced.
```html
<script src="//cx20.github.io/gltf-test/libs/filament/v1.9.11rc/filament.js"></script>
```

**Reply by cx20** [reply]
Sorry, I misunderstood your question. It seems that what you want to know is not how to specify the location of the wasm file, but how to reference the resource file using wasm.

> In windows, we have to provide the full path. 

I don't believe it has to be a full path, even in Windows. I believe you can refer to the resource by specifying a relative path.

**Reply by amarav** [reply]
@cx20 Thanks for your inputs

In the below snippet, i pass base path in resourcepath.Internally we only have relative path. So, basepath + relative we get full path to the fopen which is in another cpp file... But this doesnt seem to work :( I was trying to refer how Filament handles this but couldnt get clarity. Could you guide?

int main()
{
    std::string fl_rpath(pwd);
    char gltfpath[MAX_PATH];
    char resourcepath[MAX_PATH];
    strncpy(gltfpath, fl_rpath.c_str(), (MAX_PATH - 1));
    strncpy(resourcepath, gltfpath, (MAX_PATH - 1));
    strncat(resourcepath, "/rdf_cpp/resources/", (MAX_PATH - 1));

	
    rendercore_engine::rdf_scg_project rdf_scg_project_((rendercore_engine::rdf_project*)&rendercore_engine::mdl_project, resourcepath);
    rdf_scg_project_.Load(rendercore_engine::MdlAssetLoadMode::Synchronous);
    rendercore_engine::gfx3d_scg_display * fl_dislay0 = rdf_scg_project_.GetDisplay(0);
   
    loop = [&]
    {
        
         rdf_scg_project_.Render();
    };

// This is a function from emscripten.h, it sets a C function
// as the main event loop for the calling thread:
    emscripten_set_main_loop(main_loop, 0, true);

    return EXIT_SUCCESS;
}

**Reply by cx20** [reply]
@amarav I'm sorry. I seem to have misunderstood what you want to do.

Filament is a cross-platform library, which means that you can create native executables for each environment.
The way I presented it is for WebGL and to run in a browser.

If you want to create an application in the C++ language in a Windows environment, the following sample may be helpful.
https://github.com/google/filament/tree/main/samples

See below for instructions on how to build the sample.
https://github.com/google/filament/blob/main/BUILDING.md#windows

**Reply by amarav** [reply]
@cx20 No No.. You were right in guiding me. It is for WebGL and to run in a browser.Only thing is I need C++ language in a Windows environment to be turned into wasm and make it run in a browser. I want to port an application using another rendercore engine to be rendered in the browser

**Reply by cx20** [reply]
@amarav  I still don't think I understand exactly what you want to do.

> Only thing is I need C++ language in a Windows environment to be turned into wasm and make it run in a browser. 

What you want to achieve is to run an application written in C++ in a browser using Filament?

**Reply by amarav** [reply]
@cx20 Sorry for the confusion, I'm trying to use filament as a reference program. 
What i want is - I have a rendering engine in C++. And i want to make it run in a browser just like filament. 
The rendering engine which i use has resources to load just like filament.. So need to understand how filament works

**Reply by cx20** [reply]
@amarav Thank you for explaining it to me. I'm starting to get an idea of what you want to do.

I am not familiar with how to call `WebGL` in C++, but I think `Emscripten`'s sample code may be helpful.
https://github.com/emscripten-core/emscripten/tree/master/tests

**Reply by amarav** [reply]
Sure, thanks, I will have a look at it

**Reply by cx20** [reply]
BTW, I have created some WebGL samples using Emscripten before. I hope this is useful for you in some way.
https://github.com/cx20/webgl-test
![image](https://user-images.githubusercontent.com/1833979/107124545-e8f47100-68e7-11eb-9c74-471793915447.png)

**Reply by amarav** [reply]
Oh, thats great, will definitely be helpful. May I know if you have used opengl es2 ?

**Reply by amarav** [reply]
Also, Could you share any sample for file i/o operations in webgl? like how to read a file... if you have any by chance?

**Reply by amarav** [reply]
Also, may i know why you have used USE_GLFW=3 for compilation in

 https://github.com/cx20/webgl-test/blob/master/examples/wasm_glfw/triangle/main.cpp

**Reply by cx20** [reply]
> May I know if you have used opengl es2 ?

I don't know much about OpenGL, but looking at the `#include` in the sample code, it looks like it uses `OpenGL ES2`.
https://github.com/cx20/webgl-test/blob/master/examples/wasm/triangle/main.cpp
https://github.com/cx20/webgl-test/blob/master/examples/wasm_glfw/triangle/main.cpp
https://github.com/cx20/webgl-test/blob/master/examples/wasm_sdl/triangle/main.cpp

**Reply by cx20** [reply]
> Also, may i know why you have used USE_GLFW=3 for compilation in

What does that mean?
I think I'm using that parameter to try to use the latest version of GLFW.

**Reply by amarav** [reply]
ok. Thank you so much! 
Sorry for asking too many questions.
Is there a way to debug wasm / emscripten ?

**Reply by cx20** [reply]
I'm sorry, I've only tried Hello World in Emscripten, so I don't know how to debug it.

I just googled it, but the following may be helpful.
https://emscripten.org/docs/porting/Debugging.html

**Reply by amarav** [reply]
Thanks for your time and patience

**Comment by amarav** [comment]
Thanks @cx20
